src/amltk/optimization/optimizers/smac.py

```diff
@@ -249,13 +249,31 @@ def create(
         )
         return cls(facade=facade, fidelities=fidelities, bucket=bucket, metrics=metrics)
 
+    @overload
+    def ask(self, n: int) -> Iterable[Trial[SMACTrialInfo]]:
+        ...
+
+    @overload
+    def ask(self, n: None = None) -> Trial[SMACTrialInfo]:
+        ...
+
     @override
-    def ask(self) -> Trial[SMACTrialInfo]:
+    def ask(
+        self,
+        n: int | None = None,
```

Comment:
Mhm, I think it might be unintuitive to make the default `None`. I would rather have it be `1`. 
I know you liked had to do it this way for the recursion to work but maybe it would be cleaner to just use a new `_ask` function for that and make the public-facing interface use `1 <= n <= +inf`.

Reply:
The reason was actually for typing. If you use an `int`, you get a `list` of that size, otherwise you get a single trial.

```python
def f(n: int = 1) -> str | list[str]:
    if n == 1:
		return "hello"
	return ["hello"] * n

n = random.integer()
for message in f(n):
    print(message)  # If you get unlucky and have `1`, this would print 'h', 'e', 'l', 'l', 'o'
```

The `None` thing is basically saying that you get a single item, if you pass an integer, you get a list of that size.

Typing becomes a nightmare if you try to do it based on `Literal[1]`.

---

tldr; I won't do this change as it can cause unpredictable behaviour and doesn't offer much benefit